### PR TITLE
bpo-33712: Create the od_fast_nodes cache in OrderedDict only if needed.

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
https://bugs.python.org/issue33712

### CODE REVIEW

**Comment by A3880036** [comment]
Accept

File: Objects/odictobject.c [hunk 1]
```diff
@@ -684,7 +673,8 @@ _odict_add_new_node(PyODictObject *od, PyObject *key, Py_hash_t hash)
         Py_DECREF(key);
         return -1;
     }
-    else if (od->od_fast_nodes[i] != NULL) {
+    assert(od->od_fast_nodes != NULL);
+    if (od->od_fast_nodes[i] != NULL) {
```
Comment (A3880036) [inline Objects/odictobject.c]: Accept

File: Lib/test/test_ordered_dict.py [hunk 2]
```diff
@@ -697,9 +697,9 @@ def test_sizeof_exact(self):
         nodesize = calcsize('Pn2P')
 
         od = OrderedDict()
-        check(od, basicsize + 8*p + 8 + 5*entrysize)  # 8byte indices + 8*2//3 * entry table
+        check(od, basicsize + 8 + 5*entrysize)  # 8byte indices + 8*2//3 * entry table
         od.x = 1
```
Comment (A3880036) [inline Lib/test/test_ordered_dict.py]: Accept

File: Objects/odictobject.c [hunk 3]
```diff
@@ -585,7 +572,7 @@ _odict_resize(PyODictObject *od) {
     }
 
     /* Replace the old fast nodes table. */
-    _odict_free_fast_nodes(od);
+    PyMem_FREE(od->od_fast_nodes);
     od->od_fast_nodes = fast_nodes;
```
Comment (A3880036) [inline Objects/odictobject.c]: Accept

File: Objects/odictobject.c [hunk 4]
```diff
@@ -1636,8 +1603,9 @@ PyTypeObject PyODict_Type = {
  */
 
 PyObject *
-PyODict_New(void) {
-    return odict_new(&PyODict_Type, NULL, NULL);
+PyODict_New(void)
+{
+    return PyDict_Type.tp_new(&PyODict_Type, NULL, NULL);
```
Comment (A3880036) [inline Objects/odictobject.c]: Accept All

**Review by methane — APPROVED** [review]

**Review by vstinner — COMMENTED** [review]
PyMem_FREE should be replaced with PyMem_Free.

PyMem_FREE is just an old alias to PyMem_Free.

**Comment by serhiy-storchaka** [comment]
This is for consistency with other PyMem_FREE. Do you suggest to replace other PyMem_MALLOC, PyMem_NEW and PyMem_FREE?

**Comment by vstinner** [comment]
> Do you suggest to replace other PyMem_MALLOC, PyMem_NEW and PyMem_FREE?

I'm not sure that it's worth it to replace them. I just suggest to avoid old aliases for new code.

> This is for consistency with other PyMem_FREE.

It's up to you. It was just a remark.

**Comment by vstinner** [comment]
I merged the PR.

> vstinner removed needs backport to 3.6 needs backport to 3.7 labels a minute ago

I removed them because I understood that this change is an optimization... but oops, when I looked at https://bugs.python.org/issue33712 it looks more as a bugfix than an optimization.

@serhiy-storchaka: do you want to backport this change manually?

**Pull request merged**